### PR TITLE
Add XAML source location information to more elements - this is to aid debugging

### PR DESCRIPTION
**Pull request by mrlacey**
GitHub Issue (If applicable): n/a

This PR replaces #14645

## PR Type

What kind of change does this PR introduce?

**Other: increasing information available while debugging.**

## What is the current behavior?

XAML Source information while debugging is limited to FrameworkElements.

## What is the new behavior?

Now adds a WeakAttachedProperty to Styles, FrameworkTemplates, and VisualState related elements while debugging.

Of note:
- A new interface (`IXamlLocation`) has been added to encapsulate the common properties shared by XamlObjectDefinition and XamlMemberDefinition.
- Location is now tracked as part of `Scope` as not always the same as the current object/owner.
- `DataTemplate`s have additional complexity as what constitutes their content is treated differently depending on whether they have an `x:Key` defined.
- Previously added code trying to set `"ResourceSourceLocations"` has been removed as it didn't work as desired, and having a single property name makes the consuming code simpler.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at bebfb39</samp>

This pull request enhances the `XamlCodeGenerator` to generate and set the original source location information for the XAML elements that are created by the generator. This information is used by the hot reload feature to map the changes in the XAML file to the corresponding elements in the UI. The pull request also adds unit tests to verify the generated code and refactors some existing code to improve readability and maintainability. The main files affected are `XamlFileGenerator.cs`, `NameScope.cs`, `Subclass.cs`, `CSGenerator.cs`, `Given_HotReloadEnabledInBuild.cs`, `IXamlLocation.cs`, `XamlMemberDefinition.cs`, and `XamlObjectDefinition.cs`.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes 🤞
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Tested as far as I know how. Low existing coverage for this means the wider impact is had to tell. Creating as a draft PR to see if this shows any other issues when tests are run as part of CI.  🤞
I cannot run some tests locally, and the error messages do not include any useful information. I'm overlooking these for now as they may be related to my local configuration.

Internal Issue (If applicable):

🔥🎨461

### CODE REVIEW

**Comment by mrlacey** [comment]
UI test failures are in failing to upload a file that was (presumably) supposed to be generated by the test, not in the test itself.
If the test was supposed to produce a file but didn't, why is this not treated as the test failing?

File: src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Uno.UI.SourceGenerators.Tests.csproj [hunk 1]
```diff
@@ -293,6 +293,7 @@
 		<Compile Remove="XamlCodeGeneratorTests\Given_AttachedPropertyWithNoFuzzyMatching.cs" />
 		<Compile Remove="XamlCodeGeneratorTests\Given_Binding.cs" />
 		<Compile Remove="XamlCodeGeneratorTests\Given_Grid.cs" />
+		<Compile Remove="XamlCodeGeneratorTests\Given_Given_HotReloadEnabledInBuild.cs" />
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Uno.UI.SourceGenerators.Tests.csproj:296]: Looks like there is an extra `Given_` here

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 2]
```diff
@@ -949,6 +955,18 @@ private void BuildChildSubclasses(IIndentedStringBuilder writer, bool isTopLevel
...
 		}
 
+		/// <summary>
+		/// The "OriginalSourceLocation" can be used like DebugParseContext (set via SetBaseUri) but for elements that aren't FrameworkElements
+		/// </summary>
+		private void TrySetOriginalSourceLocation(IIndentedStringBuilder writer, string element, int lineNumber, int linePosition)
+		{
+			if (_isHotReloadEnabled)
+			{
+				//writer.AppendLineIndented($"global::Uno.UI.Helpers.MarkupHelper.SetElementProperty(__rootInstance, \"OriginalSourceLocation\", \"file:///{_fileDefinition.FilePath.Replace("\\", "/")}#L{CurrentScope.Location.LineNumber}:{CurrentScope.Location.LinePosition}\");");
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:965]: Is this needed?

**Review by jeromelaban — APPROVED** [review]
🚀

**Pull request merged**